File: 완전탐색/[완전탐색] 영선/카펫.swift
//
//  카펫.swift
//  Algorithm
//
//  Created by 임영선 on 2023/10/12.
//

import Foundation

/*
 1. 노랑블록 + 브라운블록 총 개수를 구한다.
 2. 총 개수의 곱하기 짝을 찾는다.
 3. (가로*2) + (세로*2) - 4 가 브라운 개수랑 같은 짝이 답
 
 */

func solution(_ brown:Int, _ yellow:Int) -> [Int] {
   let sum = brown+yellow
    
    // 곱하기짝을 담을 배열
    var sizes: [[Int]] = []
    
    for i in 1...sum {
        if sum % i == 0 {
            let a = sum/i
            let b = sum/a
            
            // 짝 중복 제거
            if a < b { break }
            sizes.append([a,b])
        }
    }
    
    for size in sizes {
        let brownCount = (size[0]*2 + size[1]*2)-4
        if brown == brownCount {
            return size
        }
    }

    return []
}

print(solution(24, 24))

File: 완전탐색/지후 [완전탐색]/카펫.js
function solution(brown, yellow) {

    // brown 이나 yellow가 없을 때 빈 배열 return
    if (!brown || !yellow) return [];

    // 정사각형의 가로 길이를 i로 반복하며 검사
    for (let i = 3; i <= (brown + yellow) / i; i++) {
        // 정사각형의 가로 길이 i에 따른 세로 길이 계산
        let width = Math.floor((brown + yellow) / i);

        // 주어진 노란색 타일의 개수와 비교하여 조건을 만족하는지 확인
        if ((width - 2) * (i - 2) === yellow) {
            // 조건을 만족하는 경우 가로와 세로 길이를 반환
            return [width, i];
        }
    }
}
